### PR TITLE
docs: Decision docs for start of rewrite

### PR DESCRIPTION
**Pull request by nzakas**
#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request?

Formally documents decisions around this repo.

#### What changes did you make? (Give an overview)

Added decision documents covering the following:

- Added a template
- To do a core rewrite
- Using JS/JSDoc/tsc instead of TypeScript
- Using ESM

#### Is there anything you'd like reviewers to focus on?

I did my best to piece back together the logic from our various discussions, but do let me know if any important details are missing.

### CODE REVIEW

File: decisions/003-esm.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+For all of ESLint's history, packages have been written and published in CommonJS, only sometimes providing an ESM entrypoint.
+
+## Decision
+
+All packages in the rewrite repository will be written in ESM and preferentially provide ESM entrypoints (where applicable). When possible, the packages will provide CommonJS entrypoints as well.
+
+## Consequences
+
+We will not be dogfood CommonJS-specific functionality in ESLint, so there is the possibility that some errors could creep in. We'll need to rely on our tests to catch these.
```
Comment (fasttime) [inline decisions/003-esm.md:19]: A typo:

```suggestion
We will not be dogfooding CommonJS-specific functionality in ESLint, so there is the possibility that some errors could creep in. We'll need to rely on our tests to catch these.
```

File: decisions/001-rewrite-core.md
Comment (fasttime) [inline decisions/001-rewrite-core.md:1]: Maybe worth linking to the [original discussion](https://github.com/eslint/eslint/discussions/16557) somewhere for the benfit of new readers?

**Review by fasttime — DISMISSED** [review]
LGTM, thanks. Would like @mdjermanovic to verify.

File: decisions/002-javascript-jsdoc-tsc.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+Date: 2022-10-31
+
+Status: accepted
+
+## Context
+
+The [decision to rewrite the core](./001-rewrite-core.md) in a new repository brings with it the opportunity to do things differently, which naturally brought up the idea of rewriting the core in TypeScript. While TypeScript has advantages for many projects, the nature of ESLint makes this less appealing.
+
+The ESLint project is more than just the `eslint` CLI tool. It is also [Espree](https://github.com/eslint/espree), [`eslint-scope`](https://github.com/eslint/eslint-scope), and other utilities that ESLint uses to lint JavaScript code. Part of maintaining stability in ESLint is ESLint's ability to effectively test all of these utilities every time ESLint is run. Rewriting in TypeScript would necessarily mean switching to use [`typescript-eslint`](https://typescript-eslint.io) for linting out own project, which would mean we'd no longer be dogfooding our own utilities.
```
Comment (mdjermanovic) [inline decisions/002-javascript-jsdoc-tsc.md:11]:
```suggestion
The ESLint project is more than just the `eslint` CLI tool. It is also [Espree](https://github.com/eslint/espree), [`eslint-scope`](https://github.com/eslint/eslint-scope), and other utilities that ESLint uses to lint JavaScript code. Part of maintaining stability in ESLint is ESLint's ability to effectively test all of these utilities every time ESLint is run. Rewriting in TypeScript would necessarily mean switching to use [`typescript-eslint`](https://typescript-eslint.io) for linting our own project, which would mean we'd no longer be dogfooding our own utilities.
```

File: decisions/002-javascript-jsdoc-tsc.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## Decision
+
+The rewrite will use JavaScript and JSDoc comments along with `tsc` to enforce type checking. This allows ESLint to continue to dogfood its own parser, scope analyzer, and related tools without foregoing the type safety that TypeScript provides. We will use TypeScript for defining interfaces where necessary, as this is more convenient than JSDoc format, but not for functionality.
+
+Additionally, each package in the rewrite repository will publish their own types.
```
Comment (mdjermanovic) [inline decisions/002-javascript-jsdoc-tsc.md:17]:
```suggestion
Additionally, each package in the rewrite repository will publish its own types.
```

**Review by mdjermanovic — COMMENTED** [review]
LGTM, just two possible typos.

**Pull request merged**